CEP-37 on Trunk

jaydeepkumar1984 opened the pull request:
CEP:
https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-37+Apache+Cassandra+Unified+Repair+Solution

Design doc:
https://docs.google.com/document/d/1CJWxjEi-mBABPMZ3VWJ9w5KavWfJETAGxfUpsViPcPo/edit?tab=t.0#heading=h.r112r46toau0


The Cassandra JIRA: https://issues.apache.org/jira/browse/CASSANDRA-19918

**Code review comments**

- The comments mentioned in this PR are not exhaustive; a lot of review happened on 4.1 PR, so please refer to those comments as well here: https://github.com/apache/cassandra/pull/3367/
- There have been significant code reviews happend on individual sub-PRs, some of them are here:
1) AutoRepair documentation improvements: https://github.com/jaydeepkumar1984/cassandra/pull/22
2) Implement preview_repaired auto-repair type: https://github.com/jaydeepkumar1984/cassandra/pull/4
3) RangeRepairSplitter prototype: https://github.com/jaydeepkumar1984/cassandra/pull/3
4) Make RepairTokenRangeSplitter the default implementation: https://github.com/jaydeepkumar1984/cassandra/pull/18
5) Update RepairTokenRangeSplitter to work with BTI-formatted SSTables https://github.com/jaydeepkumar1984/cassandra/pull/36
6) Deprioritize repair on nodes whose replicas are actively being repaired https://github.com/jaydeepkumar1984/cassandra/pull/51
7) Update the documentation https://github.com/jaydeepkumar1984/cassandra/pull/47
8) Avoid micro splitting of token ranges, specifically with vnodes https://github.com/jaydeepkumar1984/cassandra/pull/46
9) Avoid selecting nodes when their replicas are already scheduling repairs https://github.com/jaydeepkumar1984/cassandra/pull/51/files

jaydeepkumar1984 committed:
config description

kzalys committed:
Implement repair retries for auto-repair scheduler

kzalys committed:
Extend nodetool sstablerepairedset to affect all keyspaces when no keyspace is provided in args

jaydeepkumar1984 committed:
Make repair session timeout configurable

jaydeepkumar1984 committed:
Comments from Joshua M

jaydeepkumar1984 committed:
AutoRepair scheduler InJvm dtest

jaydeepkumar1984 committed:
Increase CircleCI parallelism 1-->4

jaydeepkumar1984 committed:
Fix a test case failure org.apache.cassandra.schema.SchemaKeyspaceTest::testConversionsInverses

jaydeepkumar1984 committed:
Trigger automatic CircleCI jobs

jaydeepkumar1984 committed:
CircleCI ask for a manual approval

kzalys committed:
Fix the JMXStandardsTest::interfaces

jaydeep1984 committed:
Circle AutoStart J8 tests

tolbertam committed:
Update IAutoRepairTokenRangeSplitter API to be more flexible

Updates IAutoRepairTokenRangeSplitter API to take in a set of tables
instead of a single table and return "RepairAssignments".

This allows the API more control over how the repairs are issued by
the AutoRepair framework, by allowing grouping of repairs by specific
tables or a single repair at the keyspace level.

Also updates IAutoRepairTokenRangeSplitter to be a ParameterizedClass,
so custom splitters can have their own configuration.

Update AutoRepair to account for these changes.

Update DefaultAutoRepairTokenSplitter to account for the change, and
to also evaluate on getRepairByKeyspace option.  This also improves
the filtering of tables that should not be repaired where previously
getRepairByKeyspace would not factor this in.

Move logic to split token ranges evenly into AutoRepairUtils so it
can be reutilized by other splitters.

kzalys committed:
Allow IR to run with MVs and CDC if write path replay is disabled

kzalys committed:
Add a nodetool command to get the nodes going through repair

jaydeepkumar1984 committed:
Adjust a couple of default configuration values

jaydeepkumar1984 committed:
Rename EMPTY --> NONE for Nodetool autorepairstatus command

jaydeepkumar1984 committed:
Minor code refactoring to simplify the repair API by removing the SLEEP_IF_REPAIR_FINISHES_QUICKLY

jaydeepkumar1984 committed:
Adjust AutoRepair on 5.0

jaydeepkumar1984 committed:
Fix unit test cases DescribeStatementTest & SchemaCQLHelperTest

jaydeepkumar1984 committed:
Safeguard the setup call to ensure it it done only once

jaydeepkumar1984 committed:
Fix InJvm test AutoRepairSchedulerTest

jaydeepkumar1984 committed:
Unit test failures AutoRepairTest.java

jaydeepkumar1984 committed:
5.0 formatting

jaydeepkumar1984 committed:
Fix flaky AutoRepairTest::testSetup and AutoRepairTest::testSafeGuardSetupCall

kzalys committed:
Fix a test failure inside pylib/cqlshlib/test/test_cqlsh_output.py

jaydeepkumar1984 committed:
AutoRepair nodetool throws an exception if the framework is disabled

jaydeepkumar1984 committed:
Run repair on bootstrapped if configured to do so

jaydeepkumar1984 committed:
Fix broken AutoRepairSchedulerTest.java & AutoRepairStatusTest.java

jaydeepkumar1984 committed:
Comments from Chris Lohfink: nix automated_ and a typo

jaydeepkumar1984 committed:
Comments from Chris Lohfink: System::currentTimeMillis --> Clock.Global::currentTimeMillis

jaydeepkumar1984 committed:
Fix a corner-case bug NPE org.apache.cassandra.repair.autorepair.AutoRepairUtils.myTurnToRunRepair

jaydeepkumar1984 committed:
Update documentation for the newly added metric

jaydeepkumar1984 committed:
Fix the failed StorageServiceServerTest test cases

jaydeepkumar1984 committed:
Formatting: Add a new line towards the end SystemDistributedKeyspace.java

jaydeepkumar1984 committed:
Formatting: Add a new line towards the end IAutoRepairTokenRangeSplitter.java

jaydeepkumar1984 committed:
Formatting: YamlConfigurationLoader.java & CassandraStreamReceiver.java

jaydeepkumar1984 committed:
Revert .circleci/config.yaml settings

michaelsembwever commented:
@masokol , @emolsson , @itskarlsson , @tommystendahl , @etedpet , @jwaeab , @VictorCavichioli , @SajidRiaz138 , @ch1bbe , @ArcturusMengsk , @DanielwEriksson , @manmagic3 

as contributors to https://github.com/Ericsson/ecchronos we would very much appreciate any last-minute-pre-CEP-vote review to this PR for Cassandra's new repair solution.   (more technical/code review will continue post-CEP vote.)

CEP:
https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-37+Apache+Cassandra+Unified+Repair+Solution

Design doc: 
https://docs.google.com/document/d/1CJWxjEi-mBABPMZ3VWJ9w5KavWfJETAGxfUpsViPcPo/edit?tab=t.0#heading=h.r112r46toau0

masokol commented:
Hi,

I think this looks promising! I have a few points:
* Combine ranges instead of splitting - in ecChronos we saw huge improvements in some scenarios (when the data is low or empty tables) compared to repairing 1 vnode at a time. This improvement scaled with amount of vnodes, although it might've been related to overhead due to running repairs through JMX. Not sure but might be worth investigating.
* Major versions, during major version upgrades like 3 -> 4 we weren't supposed to run repairs. If Cassandra plans to keep this then it would be nice for repairs to automatically pause during major version upgrades.
* Observabliity, i saw there're metrics but it would also be nice to see repair status with nodetool.
* Repair priority per table, not per node.

jaydeepkumar1984 commented:
Thanks, @masokol, for the review! Please find my response here:

> - Combine ranges instead of splitting - in ecChronos we saw huge improvements in some scenarios (when the data is low or empty tables) compared to repairing 1 vnode at a time. This improvement scaled with amount of vnodes, although it might've been related to overhead due to running repairs through JMX. Not sure but might be worth investigating.

1. Generally, empty tables or tables with a small amount of data runs through pretty fast, in seconds, so it is not a major issue for smaller/empty tables.
2. The current framework already has support to combine ranges through a setting ([src/java/org/apache/cassandra/repair/autorepair/AutoRepair.java:266](https://github.com/apache/cassandra/pull/3598/files#diff-b9a857a0f20a2f239d5e27511e4db19c87704faccb17b5205b5406e69ea08f52R266))

> - Major versions, during major version upgrades like 3 -> 4 we weren't supposed to run repairs. If Cassandra plans to keep this then it would be nice for repairs to automatically pause during major version upgrades.

This is a great suggestion, and the framework can be enhanced easily - I just filed a new sub-ticket [CASSANDRA-20013](https://issues.apache.org/jira/browse/CASSANDRA-20013) to track this 

> - Observabliity, i saw there're metrics but it would also be nice to see repair status with nodetool.

There is already a new nodetool command that would print the current status. Please take a look at it [here](https://docs.google.com/document/d/1CJWxjEi-mBABPMZ3VWJ9w5KavWfJETAGxfUpsViPcPo/edit?tab=t.0#bookmark=id.887cdlihznq7).

> - Repair priority per table, not per node.

Currently, it will repair the tables randomly, but it can be enhanced to add a priority as a CQL table property that an end user can configure, which can also be enhanced easily. Just added this enhancement to the ticket mentioned above.

jaydeepkumar1984 committed:
Code formatting: Remove Group from naming

jaydeepkumar1984 committed:
Enable repair scheduler through nodetool

jaydeepkumar1984 committed:
Increment system_distributed GENERATION

tolbertam committed:
UnrepairedBytesBasedTokenRangeSplitter prototype

Adds prototype UnrepairedBytesBasedTokenRangeSplitter that
demonstrates consuming splitter-based properties to limit the
repair assignments returned to be bound to the requested
amount of work.

jaydeepkumar1984 committed:
Option to disable write path during streaming for MV enabled tables; similar to CASSANDRA-17666

jaydeepkumar1984 committed:
cr from tolbertam - replace MV/CDC repair replay JVM properties to yaml configs

jaydeepkumar1984 committed:
ant checkstyle: toLowerCase-->toLowerCaseLocalized

jaydeepkumar1984 committed:
code formatting: DatabaseDescriptor.java

kzalys committed:
Implement preview repaired metrics

clohfink committed:
Add Repair range splitter

clohfink committed:
add assignment priorirty to table options

clohfink committed:
reorder with DefaultAutoRepairTokenSplitter

clohfink committed:
add tests for partition count estimation

tolbertam committed:
Add license header to RepairRangeSplitter

tolbertam committed:
Simplify access of repair_type_overrides, key by String

A lot of care existed to ensure each RepairType was populated in
repair_type_overrides (using ensureOverrides).

This could be simplified quite a bit by protecting access to
repair_type_overrides by enclosing it in a getOptions method that
ensures the provided RepairType has a populated entry.

Additionally, change the key type to String and use ConcurrentMap;
otherwise the change to make the enum names all caps does not work.

tolbertam committed:
Allow partial repair of range, factor in total bytes for IR

* Rename maxBytesPerScheduleBytes to maxBytesPerSchedule
* Allow partial repair of range if the smallest subrange we can repair
  is less than the remaining bytes we can repair.
* Consider totalSize for IR (because of anticompaction) sizeinRange for
  full repair.
* Increased logging while debugging.

clohfink committed:
update warnings based on repair type

jaydeepkumar1984 committed:
Rename AutoRepair --> UnifiedRepair

jaydeep1984 committed:
Revert "Rename AutoRepair --> UnifiedRepair"

This reverts commit f9f6971336c26e1bcae4110f5e0926b3e92b4565.

kzalys committed:
Improve preview repair metrics

kzalys committed:
Update src/java/org/apache/cassandra/repair/autorepair/AutoRepairConfig.java

Co-authored-by: Andrew Tolbert <6889771+tolbertam@users.noreply.github.com>

tolbertam committed:
Move size comparison up so priority can be factored in

Rather than skipping assignments based on size as they
are generated, generate all assignments.  This way
we can use table priorities as a guide for which
assignments to exclude.

There is still some more work to do here, as we
are still splitting by max bytes, where we should
really split by remaining bytes.

Some further refactoring can be done here but
I think this will put more emphasis on table
priority when selecting repair assignments
and the size of the repair is greater than
max schedule bytes.

tolbertam committed:
Various cleanup

tolbertam committed:
Handle EmpySSTableScanner being returned from getScanner

If no SSTables cover a requested range an EmptySSTableScanner
will be returned, handle this.

tolbertam committed:
Minor refactoring around bytesPerSubrange, logging

Avoid intermediate variable by making DEFAULT_SUBRANGE_SIZE
a long.

Log bytes using stringifyFileSize

tolbertam committed:
Cleanup logging of repair assignments

tolbertam committed:
Remove UnrepairedBytesBasedTokenRangeSplitter

tolbertam committed:
Refactor RepairRangeSplitter interface to accept all keyspaces

This enables us to evaluate all RepairAssignments in the schedule at once which enhances the ability to order and restrict the RepairAssignments returned based on the splitter implementation.

For example, RepairRangeSplitter's max_bytes_per_schedule would be more useful if it was evaluated for all keyspaces collectively instead of per keyspace.

Also refactors AutoRepair to order keyspaces by highest table priority.  This may be enhanced later to bucket repair submissions by priority instead of ordering keyspaces by highest table priority.

tolbertam committed:
Naming consistency

getRepairAssignmentsByKeyspace -> getRepairAssignmentsForKeyspace
getRepairAssignments -> getRepairAssignmentsForKeyspace

tolbertam committed:
Use memtable size for memtable only

tolbertam committed:
Fix dtest

tolbertam committed:
Refactor IAutoRepairTokenRangeSplitter to be Iterator based

Refactors IAutoRepairTokenRangeSplitter interface to be iterator
based.

This makes generation of RepairAssignments for priority buckets
and keyspaces lazy instead of generating the RepairAssignments all at
once.  The main value here is that repairs can take some time, and it
is beneficial to calculate repair assignments as close to when repairs
would be submitted for a keyspace as possible.

Additionally, introduced "PrioritizedRepairPlan" and
"KeyspaceRepairPlan" types.  Rather than passing around multiple-depth
Maps where the contents aren't clear, create specialized types.

Introduces PrioritizedRepairPlan.build for organizing a desired repair
plan by grouping tables together that share the same priority.

Moves RepairAssignment up to its own class to simplify the
IAutoRepairTokenRangeSplitter.

Since grouping tables by priority is now done in AutoRepair, the
splitter implementations no longer have to reason with priority so
all of this logic has been removed from the splitter code. Instead, they
will now be given a List<PrioritizedRepairPlan> that is ordered by
priority.

tolbertam committed:
Add more priority and range splitter tests

kzalys committed:
Address comments

kzalys committed:
Fix preview repair options for system_schema ks

kzalys committed:
Implement BytesPreviewed and TokenRangesPreviewed metrics

kzalys committed:
Update test/unit/org/apache/cassandra/repair/autorepair/AutoRepairParameterizedTest.java

Co-authored-by: Andrew Tolbert <6889771+tolbertam@users.noreply.github.com>

kzalys committed:
Address comments

kzalys committed:
Update metrics docs

kzalys committed:
Update metrics

tolbertam committed:
Set thread name based on RepairType.getConfigName()

Changing to upper-case enums caused the executor name to be:

AutoRepair-Repair-FULL

instead of

AutoRepair-Repair-full

jaydeepkumar1984 committed:
Flatten the auto_repair table property

jaydeepkumar1984 committed:
Refactor a missed change during the rebase

jaydeepkumar1984 committed:
Fix ant checkstyle errors

jaydeepkumar1984 committed:
Ignore keyspaces with MetaStrategy

jaydeep1984 committed:
Adjust BigTableScanner.getScanner as per 5.1 API set

jaydeepkumar1984 committed:
Fix antcheckstyle

jaydeepkumar1984 committed:
Fix ant test-latest failures for RepairRangeSplitterTest-latest_jdk17 Tests run: 15, Failures: 7

tolbertam committed:
utest: calculate expected tables dynamically instead of a fixed value

jaydeepkumar1984 committed:
Ignore repair for system_traces keyspace

jaydeepkumar1984 committed:
Fix diff40

jaydeepkumar1984 committed:
Fix diff41

jaydeepkumar1984 committed:
CR from Andy T

kzalys committed:
Implement minimum repair task duration setting for auto-repair scheduler

kzalys committed:
Cleanup

kzalys committed:
Cleanup

jaydeepkumar1984 committed:
Convert the metrics to TableMeter

jaydeep1984 committed:
Add a global type=ColumnFamily

jaydeep1984 committed:
Remove an extra empty line

jaydeepkumar1984 committed:
Revert the cassandra-4.x.xml changes

tolbertam committed:
gauge compatible meter

kzalys committed:
Address comment

jaydeepkumar1984 committed:
Remove unused imports

jaydeepkumar1984 committed:
formatting

tolbertam committed:
Make RepairTokenRangeSplitter the default implementation

Summary of impacting changes:

* RepairTokenRangeSplitter is now the default in favor of
  FixedSplitTokenRangeSplitter.

* number_of_subranges moved from a repair_type_override/global config
  to a property of FixedSplitTokenRangeSplitter.

* [get|set]autorepairconfig usability changes

Detailed breakdown of changes:

0. IAutoRepairTokenRangeSplitter changes:

  * Move RepairType from an parameter in getRepairAssignments to
    a parameter in the constructor for implementations.  This was done
    because the RepairType is always the same for a splitter instance.

  * Add setParameter and getParameter methods which are used by
    setautorepairconfig and getautorepairconfig to dynamically update
    splitter configuration.

1. [get|set]autorepairconfig changes:

  * getautorepairconfig output now shows property names instead of
    human readable names (e.g. repair_check_interval instead of
    repair eligibility check interval). This was done to make it more
    intuitive to know what properties to use for setautorepairconfig.

  * getautorepairconfig and setautorepairconfig now support viewing
    and changing splitter properties, e.g.:

    setautorepairconfig token_range_splitter.max_bytes_per_schedule 500GiB -t full

2. RepairTokenRangeSplitter changes:

  * Renames RepairRangeSplitter to RepairTokenRangeSplitter and makes it
    the default implementation.

  * Establishes defaults for each repair type to be sensible.

  * Improve javadocs detailing the primary goal of the splitter, its
    configuration and its defaults and the justifications for using them.

  * Rename variables to be consistent with their setting names.

3. FixedSplitTokenRangeSplitter changes:

  * Renames DefaultAutoRepairTokenSplitter to
    FixedSplitTokenRangeSplitter as it is no longer the default.

  * Move number_of_subranges from a global config to a property for
    this splitter.

4. RepairAssignmentIterator

  * Refactored common code from both splitter implementations into
    RepairAssignmentIterator with the aim to reduce the amount of
    boiler plate custom splitter implementations need to implement.

5. Test changes

  * Fix AutoRepairParameterizedTest to use fixed splitter so we get a
    deterministic repair plan.

  * Allow splitter to be changed programmatically, only expect it to be
    used for tests.

  * Rename CassandraSreamReceiverTest and fix it

    Whether streaming cdc/mvs into commitlogs was previously dependent
    on system properties;  update the test to account for the new
    yaml properties.

  * Fix dtest after CASSANDRA-20160

    The introduction of repair_task_min_duration causes repairs to take
    cumulatively longer for a node than 2 minutes.  To resolve this,
    set that to 0s, and also enable repair_by_keyspace and set subranges
    to 1 to reduce the overall number of repairs.

Patch by Andy Tolbert; reviewed by ___ for CASSANDRA-20179

tolbertam committed:
JDK 8 support and test rename

tolbertam committed:
Fix newAutoRepairTokenRangeSplitter duplication

tolbertam committed:
Nit-fixing in existing code

* Move curly brackets to new line
* Remove unused config declaration in constructor

tolbertam committed:
Process feedback

Also remove redundant AutoRepairConfig.RepairType for RepairType
in AutoRepairConfig.

jaydeepkumar1984 committed:
Fix the the total expected sub-ranges calculation keeping v-nodes in mind

tolbertam committed:
Feedback and update tests

* Promote node not being present in gossip to warn
* Clean up NUMBER_OF_SUBRANGES doc
* Simplify default map parameter parsing in both splitters
* Doc cleanup and make RepairAssignmentIterator fully public
* Remove unnecessary getters and setters in Splitters
* Handle case where ColumnFamilyStore not retrievable, in this case
  return no assignments as we can assume deleted.
* Always return a RepairAssignment for a table, even if empty, in
  RepairTokenRangeSplitter as node may have missed writes
* Add set|getParameters tests
* Update AutoRepairParameterizedTest to use RepairTokenRangeSplitter
* Move no-split specific test to FixedSplitTokenRangeSplitterTest

jaydeepkumar1984 committed:
Remove unused import in RepairTokenRangeSplitterTest.java

jaydeepkumar1984 committed:
Add configuration to cassandra.yaml and cassandra_later.yaml

jaydeepkumar1984 committed:
Add a doc for AutoRepair feature

jaydeepkumar1984 committed:
cr from Andy

jaydeepkumar1984 committed:
address cr from Andy

tolbertam committed:
Make AutoRepair's repair_by_keyspace default to true instead of false

Creating repair sessions by table can create an overwhelming amount of
repairs, especially with vnodes.

If a repair assignment is too big (> 64 tables by default, or > 200GB
for full/50GB for incremental) RepairTokenRangeSplitter will already
split into multiple repair assignments.

Adjusts repair_by_keyspace to default to true.

patch by Andy Tolbert; reviewed by Jaydeep Chovatia for CASSANDRA-20232

tolbertam committed:
partitions_per_assignment to not use repair_session_max_tree_depth

Updates partitions_per_assignment to not be based on
repair_session_max_tree_depth which is deprecated.  Instead, just
use 2^20.

Also updates documentation around partitions_per_assignment and cleans
up some warnings in RepairTokeRangeSplitter.

patch by Andy Tolbert; reviewed by Jaydeep Chovatia for CASSANDRA-20231

tolbertam committed:
Fix inconsistent spacing in javadoc

frankgh committed:
AutoRepairServiceMBean should not use custom types for JMX methods

Patch by Francisco Guerrero; reviewed by TBD for CASSANDRA-20185

frankgh committed:
Address comments

frankgh committed:
Fix tests and rename autoRepairConfiguration -> getAutoRepairConfiguration in the AutoRepairServiceMBean definition

jaydeepkumar1984 committed:
Adjust based on the latest trunk

jaydeepkumar1984 committed:
Adjust based on the latest trunk

kzalys committed:
Fix race condition in auto-repair scheduler

kzalys committed:
Fix race condition in auto-repair scheduler

jaydeepkumar1984 committed:
Remove commented code; use TokenRingUtils.getPrimaryRangesForEndpoint instead of StorageService.instance.getPrimaryRanges

jaydeepkumar1984 committed:
Remove commented code; use TokenRingUtils.getPrimaryRangesForEndpoint instead of StorageService.instance.getPrimaryRanges

tolbertam commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -32,6 +32,8 @@
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Sets;
+
+import org.apache.cassandra.repair.autorepair.AutoRepairConfig;
```
nit:  move import below `AuditLogOptions` import

tolbertam commented on src/java/org/apache/cassandra/service/ActiveRepairService.java:
```diff
@@ -51,13 +51,25 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Multimap;
 
+import org.apache.cassandra.concurrent.ExecutorPlus;
```
nit: Imports should go with the other org.apache ones.

tolbertam commented on src/java/org/apache/cassandra/service/ActiveRepairService.java:
```diff
@@ -1075,6 +1097,16 @@ public void setRepairPendingCompactionRejectThreshold(int value)
         DatabaseDescriptor.setRepairPendingCompactionRejectThreshold(value);
     }
 
+    public double getIncrementalRepairDiskHeadroomRejectRatio()
+    {
+        return DatabaseDescriptor.getIncrementalRepairDiskHeadroomRejectRatio();
+    }
+
+    public void setIncrementalRepairDiskHeadroomRejectRatio(double value)
+    {
+        DatabaseDescriptor.setIncrementalRepairDiskHeadroomRejectRatio(value);
+    }
+
```
Can we add these to ActiveRepairServiceMBean so we configure this from JMX?  Was just thinking that `RepairPendingCompactionRejectThreshold` is something i've had to tweak through JMX a bunch in the past

tolbertam replied:
This was added to `ActiveRepairServiceMBean` :+1:

tolbertam commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -354,6 +356,10 @@ public MemtableOptions()
 
+    // at least 20% of disk must be unused to run incremental repair
+    // if you want to disable this feature (the recommendation is not to, but if you want to disable it for whatever reason) then set the ratio to 0.0
+    public volatile double incremental_repair_disk_headroom_reject_ratio = 0.2;
```
Just noticed this prop is missing from cassandra.yaml, can you add it?

tolbertam replied:
This is present now so resolving this.

tolbertam commented on src/java/org/apache/cassandra/config/DurationSpec.java:
```diff
@@ -41,7 +42,7 @@
  * options. (CASSANDRA-15234)
  */
-public abstract class DurationSpec
+public abstract class DurationSpec implements Serializable
```
With CASSANDRA-20185 we don't need to make this Serializable anymore I think, so would be good to change this back.

tolbertam commented on src/java/org/apache/cassandra/config/ParameterizedClass.java:
```diff
@@ -30,7 +31,7 @@
 
 @Shared(scope = SIMULATION)
-public class ParameterizedClass
+public class ParameterizedClass implements Serializable
```
same as `DurationSpec` shouldn't need to be `Serializable` anymore.

tolbertam commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamReceiver.java:
```diff
@@ -25,16 +25,19 @@
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Iterables;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
```
nit: move imports

tolbertam commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamReceiver.java:
```diff
@@ -190,11 +192,11 @@ private boolean cdcRequiresWriteCommitLog(ColumnFamilyStore cfs)
      * run through the CommitLog, so they can be archived by the CDC process on discard.
      */
-    private boolean requiresWritePath(ColumnFamilyStore cfs)
+    public boolean requiresWritePath(ColumnFamilyStore cfs)
```
Looks like this is only used externally by `CassandraStreamReceiverTest`, so suggest changing to default visibility and use `@VisibleForTesting` (which would need to be imported as well).
```suggestion
    @VisibleForTesting
    boolean requiresWritePath(ColumnFamilyStore cfs)
```

tolbertam commented on src/java/org/apache/cassandra/io/sstable/format/SSTableScanner.java:
```diff
@@ -83,6 +85,14 @@ protected SSTableScanner(S sstable,
+            addRange(sstable, Range.makeRowRange(range), boundsList);
+        return boundsList;
+    }
+
```
It looks like this code exists on 4.1, but no longer exists on trunk.   I also realized that `RepairTokenRangeScanner` needs to account for `Bti` instead of using `BigTableScanner` directly.   After looking at the code,  I think we can make this work without changing this code at all and we can make it implementation agnostic.  I'll propose something.

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java:
```diff
@@ -0,0 +1,1001 @@
+                long sstableSize = reader.bytesOnDisk();
+                totalBytes += sstableSize;
+                // get the bounds of the sstable for this range using the index file but do not actually read it.
+                List<AbstractBounds<PartitionPosition>> bounds = BigTableScanner.makeBounds(reader, Collections.singleton(tokenRange));
```
A couple issues here I realized just now.

1. `makeBounds` with a `TokenRange` was removed between 4.1 and trunk.  We also had changed visibility to public which probably shouldn't have done in hindsight.
2. `BtiTableScanner` (trie-based index) is now a thing, and we need to account for that.
3. We probably don't need to use `makeBounds` directly;  I believe the only two things it was use for was checking whether the SSTable encompassed the range  (which it should since we pre-filter), and to get the approximate bytes in the range.  In fact it looks like there's an API method on `SSTable` called `onDiskSizeForPartitionPositions` that will do this for us in an implementation agnostic way.

tolbertam commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableScanner.java:
```diff
@@ -20,6 +20,8 @@
 import java.io.IOException;
 import java.util.Iterator;
 
+import javax.annotation.Nullable;
```
I'll revert this with my change to use `onDiskSizeForPartitionPositions`

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -0,0 +1,147 @@
+    @Override
+    public Iterator<KeyspaceRepairAssignments> getRepairAssignments(boolean primaryRangeOnly, List<PrioritizedRepairPlan> repairPlans)
+    {
+        return new RepairAssignmentIterator(repairPlans) {
```
nit: paren should go on next line

tolbertam commented on test/unit/org/apache/cassandra/tools/nodetool/SSTableRepairedSetTest.java:
```diff
@@ -0,0 +1,116 @@
+    private SSTableRepairedSet cmd;
+
+    @Before
+    public void setUp() {
```
nit: parens in next line (all methods in this file)

jaydeepkumar1984 replied:
Fixed in this PR: https://github.com/jaydeepkumar1984/cassandra/pull/58/files

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitterTest.java:
```diff
@@ -0,0 +1,381 @@
+    private List<String> createAndInsertTables(int count) throws Throwable
+    {
+        List<String> tableNames = new ArrayList<>();
+        for (int i = 0; i < count; i++) {
```
nit: paren on next line

tolbertam commented on test/unit/org/apache/cassandra/service/AutoRepairServiceBasicTest.java:
```diff
@@ -0,0 +1,136 @@
+    private static AutoRepairService autoRepairService;
+    private static AutoRepairConfig config;
+
+    @Before
```
nit: paren should go on next line in (methods in file)

jaydeepkumar1984 replied:
Fixed in this PR https://github.com/jaydeepkumar1984/cassandra/pull/58/files

jaydeepkumar1984 replied:
Fixed in this PR https://github.com/jaydeepkumar1984/cassandra/pull/58/files

tolbertam commented on test/unit/org/apache/cassandra/service/AutoRepairServiceRepairTypeTest.java:
```diff
@@ -0,0 +1,79 @@
+    private AutoRepairService instance;
+
+    @Parameterized.Parameters(name = "repairType={0}")
+    public static Collection<AutoRepairConfig.RepairType> repairTypes() {
```
nit: paren should go on next line in this file

tolbertam commented on test/unit/org/apache/cassandra/service/AutoRepairServiceSetterTest.java:
```diff
@@ -0,0 +1,140 @@
+    public Function<AutoRepairConfig.RepairType, T> getter;
+
+    @Parameterized.Parameters(name = "{index}: repairType={0}, arg={1}")
+    public static Collection<Object[]> testCases() {
```
nit: paren should go on next line in this file

tolbertam commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -1113,6 +1117,18 @@ public void doAuthSetup(boolean async)
+        AutoRepairService.setup();
+        if (DatabaseDescriptor.getAutoRepairConfig().isAutoRepairSchedulingEnabled())
+        {
+            logger.info("Enable auto-repair scheduling");
```
```suggestion
            logger.info("Enabling auto-repair scheduling");
```

tolbertam commented on src/java/org/apache/cassandra/tcm/sequences/BootstrapAndJoin.java:
```diff
@@ -217,6 +218,8 @@ public SequenceState executeNext()
                             return halted();
                         }
+                        // this node might have just bootstrapped; check if we should run repair immediately
+                        AutoRepairUtils.runRepairOnNewlyBootstrappedNodeIfEnabled();
```
Maybe paranoid on my part, but could we move this towards the end, maybe in the end of the `if (finishJoiningRing)` block?  This way is something unexpected happens in this code path, it doesn't skip marking the bootstrap as complete/updating CMS.

tolbertam replied:
Same applies for `BootstrapAndReplace`/`ReplaceSameAddress`

tolbertam commented on src/java/org/apache/cassandra/utils/FBUtilities.java:
```diff
@@ -85,7 +87,6 @@
 import org.apache.cassandra.io.sstable.metadata.MetadataType;
 import org.apache.cassandra.io.util.DataOutputBuffer;
 import org.apache.cassandra.io.util.DataOutputBufferFixed;
-import org.apache.cassandra.io.util.File;
```
nit: accidental import move, can this be reverted?

tolbertam commented on test/unit/org/apache/cassandra/config/YamlConfigurationLoaderTest.java:
```diff
@@ -42,6 +42,8 @@
 import static org.apache.cassandra.config.YamlConfigurationLoader.SYSTEM_PROPERTY_PREFIX;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import org.apache.cassandra.repair.autorepair.AutoRepairConfig;
```
nit: import order

tolbertam reviewed (commented):
still in process of reviewing but thought I'd leave comments as I go since I still have a bit left.

jaydeepkumar1984 committed:
Fix review comments from Andy

jaydeepkumar1984 committed:
Fix review comments from Andy

tolbertam committed:
Update RepairTokenRangeSplitter to work with BTI-formatted SSTables

While doing a review pass on the CEP-37 PR I realized that
RepairTokenRangeSplitter could not possibly work with the new BTI
sstable format introduced in 5.0 because it explicitly checks for
BigTableReader implementations.

In addition, the way it calculates the amount of bytes a range covers
in an SSTable is big-format specific.

I noticed that a new utility method
SSTableReader.onDiskSizeForPartitionPositions was added recently in
CASSANDRA-20092 that can effectively calculate this for us in an
implementation-agnostic way.

This change updates the code to use this method, and also remove any
changes we previously made in SSTableScanner and BigTableScanner for
this.

Patch by Andy Tolbert; Reviewed by ___ for CASSANDRA-20315

tolbertam committed:
Update RepairTokenRangeSplitter to work with BTI-formatted SSTables

While doing a review pass on the CEP-37 PR I realized that
RepairTokenRangeSplitter could not possibly work with the new BTI
sstable format introduced in 5.0 because it explicitly checks for
BigTableReader implementations.

In addition, the way it calculates the amount of bytes a range covers
in an SSTable is big-format specific.

I noticed that a new utility method
SSTableReader.onDiskSizeForPartitionPositions was added recently in
CASSANDRA-20092 that can effectively calculate this for us in an
implementation-agnostic way.

This change updates the code to use this method, and also remove any
changes we previously made in SSTableScanner and BigTableScanner for
this.

Patch by Andy Tolbert; Reviewed by ___ for CASSANDRA-20315

tolbertam committed:
Avoid possible division by zero

tolbertam committed:
Avoid possible division by zero

tolbertam committed:
Test with Bti and Big format sstables

tolbertam committed:
Test with Bti and Big format sstables

tolbertam committed:
Assert correct format selected

tolbertam committed:
Assert correct format selected

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/TableParams.java:
```diff
@@ -148,7 +152,9 @@ public static Builder builder(TableParams params)
-                            .readRepair(params.readRepair);
+                            .readRepair(params.readRepair)
+                            .automatedRepair(params.autoRepair)
+        ;
```
nit: move ";" on the line above

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairConfigRepairTypeTest.java:
```diff
@@ -0,0 +1,61 @@
+ */
+
+package org.apache.cassandra.repair.autorepair;
+
```
nit: remove extra line

jaydeepkumar1984 committed:
Fix review comments from Jaydeepkumar

jaydeepkumar1984 committed:
Fix review comments from Jaydeepkumar

jaydeepkumar1984 commented on src/java/org/apache/cassandra/metrics/AutoRepairMetrics.java:
```diff
@@ -0,0 +1,186 @@
+            }
+        });
+
+
```
nit: extra line

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/AutoRepairParams.java:
```diff
@@ -0,0 +1,195 @@
+import static java.lang.String.format;
+import static org.apache.cassandra.utils.LocalizeString.toLowerCaseLocalized;
+
+public final class AutoRepairParams
```
nit: add a comment

tolbertam replied:
looks like you did :) resolving.

jaydeepkumar1984 commented on src/java/org/apache/cassandra/tools/NodeProbe.java:
```diff
@@ -2506,6 +2513,126 @@ public void abortBootstrap(String nodeId, String endpoint)
+        autoRepairProxy.setAutoRepairEnabled(repairType, enabled);
+    }
+
+    public void setRepairThreads(String repairType, int repairThreads)
```
s/setRepairThreads/setAutoRepairThreads/ here and elsewhere in this class

jaydeepkumar1984 committed:
Fix review comments from Andy

jaydeepkumar1984 committed:
Fix review comments from Andy

jaydeepkumar1984 commented on src/java/org/apache/cassandra/metrics/AutoRepairMetrics.java:
```diff
@@ -0,0 +1,185 @@
+public class AutoRepairMetrics
+{
+    public static final String TYPE_NAME = "autorepair";
+    public Gauge<Integer> repairsInProgress;
```
This and the other _Gauge_ in this class can be `final`

jaydeepkumar1984 commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairConfig.java:
```diff
@@ -0,0 +1,565 @@
+            opts.table_max_repair_time = new DurationSpec.IntSecondsBound("6h");
+            opts.materialized_view_repair_enabled = false;
+            opts.token_range_splitter = new ParameterizedClass(DEFAULT_SPLITTER.getName(), Collections.emptyMap());
+            opts.initial_scheduler_delay = new DurationSpec.IntSecondsBound("5m"); // 5 minutes
```
The comment is not needed here

jaydeepkumar1984 commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java:
```diff
@@ -0,0 +1,876 @@
+    COL_REPAIR_TYPE, COL_HOST_ID);
+
+    final static String RECORD_FINISH_REPAIR_HISTORY = String.format(
+
```
nit: remove empty line

jaydeepkumar1984 committed:
Additional review comments from Jaydeepkumar

jaydeepkumar1984 committed:
Additional review comments from Jaydeepkumar

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/AutoRepairParams.java:
```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    private ImmutableMap<String, String> options;
```
Make this a final.

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/AutoRepairParams.java:
```diff
@@ -0,0 +1,198 @@
+        return StringUtils.isNumeric(value);
+    }
+
+
```
nit: remove extra line

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/SchemaKeyspace.java:
```diff
@@ -559,7 +561,9 @@ private static void addTableParamsToRowBuilder(TableParams params, Row.SimpleBui
-               .add("extensions", params.extensions);
+               .add("extensions", params.extensions)
+               .add("auto_repair", params.autoRepair.asMap());
+
```
nit: remove the new line

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/SchemaKeyspace.java:
```diff
@@ -1042,7 +1046,9 @@ static TableParams createTableParamsFromRow(UntypedResultSet.Row row)
-                                                 .readRepair(getReadRepairStrategy(row));
+                                                 .readRepair(getReadRepairStrategy(row))
+                                                 .automatedRepair(AutoRepairParams.fromMap(row.getFrozenTextMap("auto_repair")))
+;
```
move it to the above line

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/SystemDistributedKeyspace.java:
```diff
@@ -159,6 +163,29 @@ private SystemDistributedKeyspace()
+    private static final TableMetadata AutoRepairPriorityTable =
+            parse(AUTO_REPAIR_PRIORITY, "Auto repair priority for each group", AUTO_REPAIR_PRIORITY_CQL).build();
+
+
```
nit: remove extra line

jaydeepkumar1984 commented on src/java/org/apache/cassandra/schema/TableParams.java:
```diff
@@ -69,7 +69,9 @@ public enum Option
-        READ_REPAIR;
+        READ_REPAIR,
+        AUTO_REPAIR,
+        ;
```
move it above

jaydeepkumar1984 commented on src/java/org/apache/cassandra/service/AutoRepairServiceMBean.java:
```diff
@@ -0,0 +1,74 @@
+public interface AutoRepairServiceMBean
+{
+    /**
+     * Enable or disable auto-repair for a given repair type
```
this comment is not needed as APIs are self explanatory

jaydeepkumar1984 commented on src/java/org/apache/cassandra/tools/NodeProbe.java:
```diff
@@ -177,6 +179,8 @@ public class NodeProbe implements AutoCloseable
     protected PermissionsCacheMBean pcProxy;
     protected RolesCacheMBean rcProxy;
+    protected AutoRepairServiceMBean autoRepairProxy;
+
```
nit: remove new line

jaydeepkumar1984 commented on src/java/org/apache/cassandra/utils/FBUtilities.java:
```diff
@@ -66,6 +66,7 @@
 import com.google.common.base.Suppliers;
 import com.google.common.collect.ImmutableList;
 import com.vdurmont.semver4j.Semver;
+
```
remove unwanted code change

jaydeepkumar1984 commented on test/distributed/org/apache/cassandra/distributed/test/repair/AutoRepairSchedulerTest.java:
```diff
@@ -0,0 +1,136 @@
+ */
+public class AutoRepairSchedulerTest extends TestBaseImpl
+{
+
```
nit: remove empty line

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/config/YamlConfigurationLoaderTest.java:
```diff
@@ -271,6 +273,12 @@ public void fromMapTest()
+                                                               "global_settings", ImmutableMap.of("number_of_repair_threads",
+                                                                                                  1),
+                                                               "repair_type_overrides", ImmutableMap.of(
+        "full", ImmutableMap.of("number_of_repair_threads",
```
nit: indentation

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairConfigTest.java:
```diff
@@ -0,0 +1,471 @@
+{
+    private AutoRepairConfig config;
+
+    private Set<String> testSet = ImmutableSet.of("dc1");
```
make this final

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairKeyspaceTest.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Iterator;
+import java.util.Set;
+
+import org.apache.cassandra.schema.*;
```
nit: provide complete _import_ path

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairUtilsTest.java:
```diff
@@ -0,0 +1,492 @@
+ */
+public class AutoRepairUtilsTest extends CQLTester
+{
+    static RepairType repairType = RepairType.INCREMENTAL;
```
final

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/SSTableRepairedAtTest.java:
```diff
@@ -0,0 +1,174 @@
+        table2 = Keyspace.open(TEST_KEYSPACE).getColumnFamilyStore("table2");
+        assert table2 != null;
+    }
+    @Test
```
add a new line

jaydeepkumar1984 committed:
Additional review comments from Andy

jaydeepkumar1984 committed:
Additional review comments from Andy

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairParameterizedTest.java:
```diff
@@ -0,0 +1,797 @@
+        assertEquals(Arrays.asList(MV), AutoRepairUtils.getAllMVs(repairType, keyspace, cfm));
+        config.setMaterializedViewRepairEnabled(repairType, false);
+    }
+
```
nit: remove a new line

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/config/YamlConfigurationLoaderTest.java:
```diff
@@ -289,6 +298,9 @@ public void fromMapTest()
         assertEquals(true, config.client_encryption_options.enabled); // Check a nested object
         assertEquals(new DataStorageSpec.IntBytesBound("5B"), config.internode_socket_send_buffer_size); // Check names backward compatibility (CASSANDRA-17141 and CASSANDRA-15234)
         assertEquals(new DataStorageSpec.IntBytesBound("5B"), config.internode_socket_receive_buffer_size); // Check names backward compatibility (CASSANDRA-17141 and CASSANDRA-15234)
+        assertEquals(true, config.auto_repair.enabled);
```
assertTrue

jaydeepkumar1984 committed:
Additional review comments from Jaydeepkumar

jaydeepkumar1984 committed:
Additional review comments from Jaydeepkumar

jaydeepkumar1984 commented:
There are a few nits that I have added, but overall it looks good now, I am approving this PR from my side 👍

jaydeepkumar1984 committed:
Incorporate NEWS.txt review comment

jaydeepkumar1984 committed:
Incorporate NEWS.txt review comment

jaydeepkumar1984 committed:
A new table property should be wrapped inside a config

jaydeepkumar1984 committed:
A new table property should be wrapped inside a config

jaydeepkumar1984 committed:
Skip displaying the table property if AutoRepair is disabled

jaydeepkumar1984 committed:
Skip displaying the table property if AutoRepair is disabled

jaydeepkumar1984 committed:
Fix broken tests

jaydeepkumar1984 committed:
Fix broken tests

jaydeepkumar1984 committed:
Extend DescribeStatementTest to cover properties when AutoRepair scheduler is enabled

jaydeepkumar1984 committed:
Extend DescribeStatementTest to cover properties when AutoRepair scheduler is enabled

kzalys committed:
Address comments

kzalys committed:
Address comments

kzalys committed:
Address comments

kzalys committed:
Address comments

kzalys committed:
Fix formatting

kzalys committed:
Fix formatting

tolbertam committed:
Account for vnodes with repair_by_keyspace in RepairTokenRangeSplitter

Updates RepairTokenRangeSplitter to call getRepairAssignmentsForKeyspace
for each individual token range so repair_by_keyspace grouping works
correctly.

Patch by Andy Tolbert; Reviewed by Jaydeepkumar Chovatia for CASSANDRA-20392

tolbertam committed:
Account for vnodes with repair_by_keyspace in RepairTokenRangeSplitter

Updates RepairTokenRangeSplitter to call getRepairAssignmentsForKeyspace
for each individual token range so repair_by_keyspace grouping works
correctly.

Patch by Andy Tolbert; Reviewed by Jaydeepkumar Chovatia for CASSANDRA-20392

tolbertam committed:
Make work with JDK11 (and JDK8)

tolbertam committed:
Make work with JDK11 (and JDK8)

jaydeepkumar1984 committed:
Adjust as per latest trunk as of 03-Mar-2025

jaydeepkumar1984 committed:
Adjust as per latest trunk as of 03-Mar-2025

jaydeepkumar1984 committed:
Type & Remove whitespace

jaydeepkumar1984 committed:
Type & Remove whitespace

jaydeepkumar1984 committed:
Ensure the main config obj has been initialized before checking the yaml property

jaydeepkumar1984 committed:
Ensure the main config obj has been initialized before checking the yaml property

jaydeepkumar1984 committed:
Use the last repair stat to confirm that the repair ran indeed instead of sleep to avoid flaky test

jaydeepkumar1984 committed:
Use the last repair stat to confirm that the repair ran indeed instead of sleep to avoid flaky test

jaydeepkumar1984 committed:
Incorporate the code review comments

jaydeepkumar1984 committed:
Incorporate the code review comments

jaydeepkumar1984 committed:
Incorporate the code review comments

jaydeepkumar1984 committed:
Incorporate the code review comments

jaydeepkumar1984 committed:
Group all public statements together

jaydeepkumar1984 committed:
Group all public statements together

jaydeepkumar1984 committed:
No need for try/catch in the test case

jaydeepkumar1984 committed:
No need for try/catch in the test case

jaydeepkumar1984 committed:
Avoid micros splitting of token ranges for FixedSplitTokenRangeSplitter

jaydeepkumar1984 committed:
Avoid micros splitting of token ranges for FixedSplitTokenRangeSplitter

jaydeepkumar1984 committed:
Make the number_of_subranges parameterized in test case

jaydeepkumar1984 committed:
Make the number_of_subranges parameterized in test case

tolbertam committed:
Auto Repair Documentation improvements, bytes_per_assignment to 50GiB

Updates auto repair documentation and moves it under
managing/operating/auto_repair.adoc.

Updates existing repair documentation to reference the auto repair
documentation.

Updates bytes_per_assignment default to 50GiB (from 200GiB) for all
report types to be more conservative.

Clean up some syntactial issues and formatting in cassandra.yaml files.

patch by Andy Tolbert; reviewed by ____ for CASSANDRA-20421

tolbertam committed:
Auto Repair Documentation improvements, bytes_per_assignment to 50GiB

Updates auto repair documentation and moves it under
managing/operating/auto_repair.adoc.

Updates existing repair documentation to reference the auto repair
documentation.

Updates bytes_per_assignment default to 50GiB (from 200GiB) for all
report types to be more conservative.

Clean up some syntactial issues and formatting in cassandra.yaml files.

patch by Andy Tolbert; reviewed by ____ for CASSANDRA-20421

tolbertam committed:
Revert tombstones.adoc no op changes.

tolbertam committed:
Revert tombstones.adoc no op changes.

tolbertam commented:
I a bit of time reviewing documentation and configuration today and have created CASSANDRA-20421 and https://github.com/jaydeepkumar1984/cassandra/pull/47 to improve the docs, configuration and to make the default configuration more conservative.

tolbertam commented on conf/cassandra.yaml:
```diff
@@ -2614,3 +2621,124 @@ drop_compact_storage_enabled: false
+
+# At least 20% of disk must be unused to run incremental repair. It is useful to avoid disks filling up during
+# incremental repair as anti-compaction during incremental repair may contribute to additional space temporarily.
+# if you want to disable this feature (the recommendation is not to, but if you want to disable it for whatever reason)
```
I think like `reject_repair_compaction_threshold`, we should default this to not do anything as to not change the behavior (should not reject repairs by default, users should have to opt into it).  Was thinking and I realized that there can be a possible downside of this feature, which is if you disable IR while you are low on disk, you might not be able to compact away shadowed data or tombstones because data exists in both the repaired/unrepaired set and can't be compacted together.

tolbertam replied:
Just adding another comment here in case this was missed.  I think it may be reasonable to keep this default behavior, but we should just make note of this somewhere in the NEWS entry or upgrade guide.

tolbertam commented on conf/cassandra.yaml:
```diff
@@ -2614,3 +2621,124 @@ drop_compact_storage_enabled: false
+# incremental repair as anti-compaction during incremental repair may contribute to additional space temporarily.
+# if you want to disable this feature (the recommendation is not to, but if you want to disable it for whatever reason)
+# then set the ratio to 0.0
+# incremental_repair_disk_headroom_reject_ratio = 0.2;
```
Should be
```suggestion
# incremental_repair_disk_headroom_reject_ratio: 0.2;
```

Addressed on https://github.com/jaydeepkumar1984/cassandra/pull/47

tolbertam committed:
Add documentation for reject_repair_compaction_threshold

tolbertam committed:
Add documentation for reject_repair_compaction_threshold

tolbertam committed:
Update src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update conf/cassandra.yaml

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Doc and conf fixes, initial IR enablement guidance

tolbertam committed:
Doc and conf fixes, initial IR enablement guidance

tolbertam committed:
Update src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

tolbertam committed:
Update src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

jaydeepkumar1984 committed:
Review comments from Andy

jaydeepkumar1984 committed:
Review comments from Andy

jaydeepkumar1984 committed:
Add v-nodes and non v-nodes test cases

jaydeepkumar1984 committed:
Add v-nodes and non v-nodes test cases

jaydeepkumar1984 committed:
Adjust the comment

jaydeepkumar1984 committed:
Adjust the comment

jaydeepkumar1984 committed:
Add the missing nodetool way to update repair_by_keyspace

jaydeepkumar1984 committed:
Add the missing nodetool way to update repair_by_keyspace

jaydeepkumar1984 committed:
Use common setup code for FixedSplitTokenRangeSplitterTest*.java

jaydeepkumar1984 committed:
Use common setup code for FixedSplitTokenRangeSplitterTest*.java

jaydeepkumar1984 committed:
Fix FixedSplitTokenRangeSplitterHelper.java failure

jaydeepkumar1984 committed:
Fix FixedSplitTokenRangeSplitterHelper.java failure

tolbertam committed:
Update num_of_subranges

tolbertam committed:
Update num_of_subranges

jaydeepkumar1984 committed:
Address review comments: update comment/remove dead code

jaydeepkumar1984 committed:
Address review comments: update comment/remove dead code

jaydeepkumar1984 committed:
Adjust comment for DEFAULT_NUMBER_OF_SUBRANGES and NUMBER_OF_SUBRANGES

jaydeepkumar1984 committed:
Adjust comment for DEFAULT_NUMBER_OF_SUBRANGES and NUMBER_OF_SUBRANGES

kzalys committed:
Move auto-repair retry policy config to repair-type level

kzalys committed:
Move auto-repair retry policy config to repair-type level

tolbertam committed:
Avoid returning MY_TURN when replicas are actively taking their turn

Previously AutoRepairUtils.myTurnToRunRepair would return MY_TURN
if there is capacity to select a new node for repair and the current
node has the oldest lastRepairFinishTime.

This change adjusts this method to make use of a newly added method
getMostEligibleHostToRepair which will identify the node with the
oldest lastRepairFinishTime of which none of the node's replicas
are actively repairing.

This should prevent replicas from actively repairing concurrently,
which would increase load on replica sets and also create anticompaction
conflicts for incremental repair.

Behavior is configurable by allow_parallel_replica_repair and
allow_parallel_replica_repair_across_schedules configuration.

patch by Andy Tolbert, reviewed by Jaydeepkumar Chovatia and Francisco Guerrero for CASSANDRA-20180

tolbertam committed:
Avoid returning MY_TURN when replicas are actively taking their turn

Previously AutoRepairUtils.myTurnToRunRepair would return MY_TURN
if there is capacity to select a new node for repair and the current
node has the oldest lastRepairFinishTime.

This change adjusts this method to make use of a newly added method
getMostEligibleHostToRepair which will identify the node with the
oldest lastRepairFinishTime of which none of the node's replicas
are actively repairing.

This should prevent replicas from actively repairing concurrently,
which would increase load on replica sets and also create anticompaction
conflicts for incremental repair.

Behavior is configurable by allow_parallel_replica_repair and
allow_parallel_replica_repair_across_schedules configuration.

patch by Andy Tolbert, reviewed by Jaydeepkumar Chovatia and Francisco Guerrero for CASSANDRA-20180

kzalys committed:
Update docs

kzalys committed:
Update docs

jaydeepkumar1984 committed:
Add toString to AutoRepairConfig.java

jaydeepkumar1984 committed:
Add toString to AutoRepairConfig.java

tolbertam commented on conf/cassandra.yaml:
```diff
@@ -2614,3 +2621,162 @@ drop_compact_storage_enabled: false
+#     materialized_view_repair_enabled: false
+#     # Delay before starting repairs after a node restarts to avoid repairs starting immediately after a restart.
+#     initial_scheduler_delay: 5m
+#     # Timeout for resuming stuck repair sessions.
```
A bit pendatic but we should change resuming to retrying as I think  its more concise.

```suggestion
#     # Timeout for retrying stuck repair sessions.
```

jaydeepkumar1984 replied:
Incorporated

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepair.java:
```diff
@@ -0,0 +1,577 @@
+        repairStates = new EnumMap<>(AutoRepairConfig.RepairType.class);
+        for (AutoRepairConfig.RepairType repairType : AutoRepairConfig.RepairType.values())
+        {
+            repairExecutors.put(repairType, executorFactory().scheduled(false, "AutoRepair-Repair-" + repairType.getConfigName(), Thread.NORM_PRIORITY));
```
Just noticed that we are creating a repair thread for each repair type, even if it auto repair isn't enabled.   Should we perhaps move this for loop into `setup`?

tolbertam replied:
Actually I just realized that while it creates executors, it doesn't create threads.  I still think it would be good to move this to setup though :+1:

jaydeepkumar1984 replied:
Incorporated

tolbertam commented on src/java/org/apache/cassandra/service/AutoRepairService.java:
```diff
@@ -0,0 +1,326 @@
+    public void checkCanRun(RepairType repairType)
+    {
+        if (!config.isAutoRepairSchedulingEnabled())
+            throw new ConfigurationException("Auto-repair scheduller is disabled.");
```
```suggestion
            throw new ConfigurationException("Auto-repair scheduler is disabled.");
```

jaydeepkumar1984 replied:
Incorporated

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepair.java:
```diff
@@ -0,0 +1,577 @@
+    }
+
+    // repairAsync runs a repair session of the given type asynchronously.
+    public void repairAsync(AutoRepairConfig.RepairType repairType)
```
Just noticed this is only used for tests currently.  I see that `repair` is only ever called from the `scheduleWithFixedDelay` call in set up, so is it possible this is dead code?

jaydeepkumar1984 replied:
Yes, it is a dead code; just removed it

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepair.java:
```diff
@@ -0,0 +1,577 @@
+        public void progress(String tag, ProgressEvent event)
+        {
+            ProgressEventType type = event.getType();
+            String message = String.format("[%s] %s", format.format(Clock.Global.currentTimeMillis()), event.getMessage());
```
very small nit:

```suggestion
            String message = String.format("[%s] %s", format.format(timeFunc.get()), event.getMessage());
```

jaydeepkumar1984 replied:
Incorporated

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairConfig.java:
```diff
@@ -0,0 +1,609 @@
+        return applyOverrides(repairType, opt -> opt.token_range_splitter);
+    }
+
+    /**
+     * Set a new token range splitter, this is not meant to be used other than for testing.
+     */
+    @VisibleForTesting
+    void setTokenRangeSplitter(RepairType repairType, ParameterizedClass tokenRangeSplitter)
+    {
+        getOptions(repairType).token_range_splitter = tokenRangeSplitter;
+        tokenRangeSplitters.remove(repairType);
+    }
+
```
Looks like I added this but its no longer being used, mind removing it?

jaydeepkumar1984 replied:
Removed

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterHelper.java:
```diff
@@ -0,0 +1,199 @@
+    {
+        int numberOfSplits = calcSplits(numTokens, numberOfSubRanges);
+        AutoRepairService.instance.getAutoRepairConfig().setRepairByKeyspace(repairType, false);
+        Collection<Range<Token>> tokens = StorageService.instance.getPrimaryRanges(KEYSPACE);
```
Looks like this is deprecated in CEP-21, change we change the 3 uses of this to?:

```
TokenRingUtils.getPrimaryRangesForEndpoint(KEYSPACE, FBUtilities.getBroadcastAddressAndPort());
```

jaydeepkumar1984 replied:
Incorporated

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java:
```diff
@@ -0,0 +1,1189 @@
+    "UPDATE %s.%s SET %s = %s + ?  WHERE %s = ?", SchemaConstants.DISTRIBUTED_KEYSPACE_NAME,
+    SystemDistributedKeyspace.AUTO_REPAIR_PRIORITY, COL_REPAIR_PRIORITY, COL_REPAIR_PRIORITY, COL_REPAIR_TYPE);
+
+    final static String INSERT_NEW_REPAIR_HISTORY = String.format(
+    "INSERT INTO %s.%s (%s, %s, %s, %s, %s, %s) values (?, ? ,?, ?, {}, ?) IF NOT EXISTS",
+    SchemaConstants.DISTRIBUTED_KEYSPACE_NAME, SystemDistributedKeyspace.AUTO_REPAIR_HISTORY, COL_REPAIR_TYPE,
+    COL_HOST_ID, COL_REPAIR_START_TS, COL_REPAIR_FINISH_TS, COL_DELETE_HOSTS, COL_DELETE_HOSTS_UPDATE_TIME);
+
+    final static String ADD_HOST_ID_TO_DELETE_HOSTS = String.format(
+    "UPDATE %s.%s SET %s = %s + ?, %s = ? WHERE %s = ? AND %s = ? IF EXISTS"
+    , SchemaConstants.DISTRIBUTED_KEYSPACE_NAME, SystemDistributedKeyspace.AUTO_REPAIR_HISTORY, COL_DELETE_HOSTS,
+    COL_DELETE_HOSTS, COL_DELETE_HOSTS_UPDATE_TIME, COL_REPAIR_TYPE, COL_HOST_ID);
```
I just realized we still have these two statements that use LWTs.  Was just looking and was wondering if we really still needed it?

I think arguably we should probably keep it as is; and it looks like it's rather low risk.

We only create new repair history when we detect it doesn't exist for a host (either on first start, or when a new node is added);  we only add a host id to delete hosts when we detect that a node is no longer in the ring.

jaydeepkumar1984 replied:
Yeah, this is to handle a rare corner-case scenario in that two nodes simultaneously add the same rows to the history table.

tolbertam commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -1113,6 +1121,18 @@ public void doAuthSetup(boolean async)
+            logger.info("Enabling auto-repair scheduling");
+            AutoRepair.instance.setup();
+        }
+        logger.info("AutoRepair setup complete!");
```
nit: should move this into the `if` statement above as if you have the feature disabled it makes it appear as if it was enabled.

jaydeepkumar1984 replied:
Incorporated

jaydeepkumar1984 committed:
Create AutoRepair executors only if enabled and a few other nits

jaydeepkumar1984 committed:
Create AutoRepair executors only if enabled and a few other nits

jaydeepkumar1984 committed:
parens in next line

jaydeepkumar1984 committed:
parens in next line

tolbertam reviewed (approved):
Added my tentative +1 with some test results in a [JIRA comment](https://issues.apache.org/jira/browse/CASSANDRA-19918?focusedCommentId=17941438&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17941438) with a final +1 pending review feedback from Chris and Francisco and a final sync with trunk, i'll stay involved on processing feedback and testing changes.  Really looking forward to this landing, nice work @jaydeepkumar1984 and everyone else involved!

jaydeepkumar1984 committed:
Adjust based on the latest trunk

jaydeepkumar1984 committed:
Merge branch 'trunk_cep_37_rebase_on_apr_17_2025' into trunk_cep_37

jaydeepkumar1984 commented:
Rebase on top of the latest trunk has been completed; all the unit tests are passing on CircleCI
- [java17_pre-commit_tests](https://app.circleci.com/pipelines/github/jaydeepkumar1984/cassandra/531/workflows/432995d7-efe5-4f8d-9f7e-1eef1873e612)
- [java17_separate_tests](https://app.circleci.com/pipelines/github/jaydeepkumar1984/cassandra/531/workflows/73a7a1f5-7030-4ed8-9768-48adca81f5f1)
- [java11_pre-commit_tests](https://app.circleci.com/pipelines/github/jaydeepkumar1984/cassandra/531/workflows/98cdee5e-fb46-48a6-b27a-10a36e616f78)
- [java11_separate_tests](https://app.circleci.com/pipelines/github/jaydeepkumar1984/cassandra/531/workflows/d69f7f72-2e38-4b14-b269-361e56a4ba8e)

Many dtests are failing on CircleCI due to resource limitations. For that, I have run them internally inside our infrastructure and most of them are passing as expected.

frankgh commented on src/java/org/apache/cassandra/service/ActiveRepairService.java:
```diff
@@ -1086,6 +1108,16 @@ public void setRepairPendingCompactionRejectThreshold(int value)
+
+    public void setIncrementalRepairDiskHeadroomRejectRatio(double value)
+    {
+        DatabaseDescriptor.setIncrementalRepairDiskHeadroomRejectRatio(value);
```
can we validate for only positive numbers to be set here  between the values of 0-1?

frankgh commented on src/java/org/apache/cassandra/schema/SystemDistributedKeyspace.java:
```diff
@@ -84,7 +85,7 @@ private SystemDistributedKeyspace()
      * // TODO: TCM - how do we evolve these tables?
      */
-    public static final long GENERATION = 6;
+    public static final long GENERATION = 7;
```
can we add a description for gen 7 above?

jaydeepkumar1984 replied:
Done

frankgh commented on doc/modules/cassandra/pages/managing/operating/metrics.adoc:
```diff
@@ -1081,6 +1088,67 @@ partitions processed per logged batch
+due to priority set in the automated repair
+
+|RepairDelayedByReplica |Counter |Represents occurrences of a node's turn being
+delayed because a replica was currently taking its turn. Only revelent if
```
relevant maybe?

```suggestion
delayed because a replica was currently taking its turn. Only relevant if
```

frankgh commented on doc/modules/cassandra/pages/managing/operating/metrics.adoc:
```diff
@@ -1081,6 +1088,67 @@ partitions processed per logged batch
+`allow_parallel_replica_repair` is false.
+
+|RepairDelayedBySchedule |Counter |Represents occurrences of a node's turn being
+delayed because it was already being repaired in another schedule. Only relevent
```
relevant maybe? 
```suggestion
delayed because it was already being repaired in another schedule. Only relevant
```

jaydeepkumar1984 committed:
Merge branch 'trunk' into trunk_cep_37

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java:
```diff
@@ -0,0 +1,932 @@
+ * <p>
+ * While this splitter has a lot of tuning parameters, the expectation is that the established default configuration
+ * shall be sensible for all {@link org.apache.cassandra.repair.autorepair.AutoRepairConfig.RepairType}'s. The following
+ * configuration parameters are offered.
```
did we intend to list the configuration parameters below in the javadocs?

jaydeepkumar1984 replied:
Incorporated

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java:
```diff
@@ -0,0 +1,932 @@
+    private static final Logger logger = LoggerFactory.getLogger(RepairTokenRangeSplitter.class);
+
+    // Default max bytes to 100TiB, which is much more readable than Long.MAX_VALUE
+    private static final DataStorageSpec.LongBytesBound MAX_BYTES = new DataStorageSpec.LongBytesBound(100_000, DataStorageSpec.DataStorageUnit.GIBIBYTES);
```
I think this should be 102_400 to be equal to 100TiB
```suggestion
    private static final DataStorageSpec.LongBytesBound MAX_BYTES = new DataStorageSpec.LongBytesBound(102_400, DataStorageSpec.DataStorageUnit.GIBIBYTES);
```

frankgh commented on src/java/org/apache/cassandra/schema/AutoRepairParams.java:
```diff
@@ -0,0 +1,197 @@
+        for (Map.Entry<String, String> entry : DEFAULT_OPTIONS.entrySet())
+        {
+            optionsMap.put(entry.getKey(), entry.getValue());
+        }
```
NIT, we can initialize this in the constructor directly:
```suggestion
        Map<String, String> optionsMap = new TreeMap<>(DEFAULT_OPTIONS);
```

frankgh commented on src/java/org/apache/cassandra/schema/AutoRepairParams.java:
```diff
@@ -0,0 +1,197 @@
+        String enabled = options.get(option);
+        return enabled == null
+               ? Boolean.parseBoolean(DEFAULT_OPTIONS.get(option))
+               : Boolean.parseBoolean(enabled);
```
NIT:
```suggestion
        String enabled = options.getOrDefault(option, DEFAULT_OPTIONS.get(option));
        return Boolean.parseBoolean(enabled);
```

frankgh commented on src/java/org/apache/cassandra/schema/AutoRepairParams.java:
```diff
@@ -0,0 +1,197 @@
+        String priority = options.get(Option.PRIORITY.toString());
+        return priority == null
+               ? Integer.parseInt(DEFAULT_OPTIONS.get(Option.PRIORITY.toString()))
+               : Integer.parseInt(priority);
```
NIT:
```suggestion
        String priority = options.getOrDefault(Option.PRIORITY.toString(), DEFAULT_OPTIONS.get(Option.PRIORITY.toString()));
        return Integer.parseInt(priority);
```

frankgh reviewed (commented):
Great work! I left a few minor comments. Also I have a few NITs , so feel free to ignore those

jaydeepkumar1984 committed:
Update doc/modules/cassandra/pages/managing/operating/metrics.adoc

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

jaydeepkumar1984 committed:
Update src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

jaydeepkumar1984 committed:
Update src/java/org/apache/cassandra/schema/AutoRepairParams.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

jaydeepkumar1984 committed:
Update src/java/org/apache/cassandra/schema/AutoRepairParams.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

jaydeepkumar1984 committed:
Update src/java/org/apache/cassandra/schema/AutoRepairParams.java

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>

jaydeepkumar1984 committed:
Validate disk headroom for IR and a couple of comment updates

jaydeepkumar1984 committed:
Format { on the new line

frankgh reviewed (approved):
+1

jaydeepkumar1984 committed:
Merge pull request #63 from jaydeepkumar1984/trunk_cep_37_cassandra_20572

Validate disk headroom for IR and a couple of comment updates

jaydeepkumar1984 committed:
Update doc/modules/cassandra/pages/managing/operating/metrics.adoc

Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>
